rpcclient/infrastructure.go

```diff
@@ -927,14 +934,37 @@ func (c *Client) sendCmd(ctx context.Context, cmd interface{}) chan *response {
 		return newFutureError(err)
 	}
 
-	// Generate the request and send it along with a channel to respond on.
 	responseChan := make(chan *response, 1)
+	relayChan := make(chan *response, 1)
+	// Pass results along for as long as ctx is not done.
+	go func() {
```

Comment:
This seems incredibly wasteful to me.  It will start a goroutine for every single command.  I would highly recommend sending the context along with the request.

Reply:
Do you mean to add the ctx as a field to the request struct? And then maybe check it periodically? I think that is useful up until sending, but after the message is sent, I wanted a failure as soon as the ctx is done.

Now it waits for a ctx done in receiveFuture, which doesn't need a go routine. However, the change is very far reaching in terms of lines of code.

If this looks ok I will fix the tests. Also, do you think it's necessary to remove the request from the map? If so receiveFuture will need to get a little bit smarter.

Reply:
The updates are definitely more in line with how I think it would be most ideal and efficient, but I don't think the `Receive` functions should be taking a context, because it might not be the same one the async request was started with and can lead to really unexpected behavior.

Instead, I think each future should house the context it was created with.  For example, perhaps something like this:

```Go
type FutureGenerateResult struct {
	ctx context.Context
	c   chan *response
}
func (r FutureGenerateResult) Receive() ([]*chainhash.Hash, error) {
	res, err := receiveFuture(r.ctx, r.c)
```

That way, the caller doesn't change and it can't lead to unexpected behavior.  e.g:
```Go
ctx := context.WithTimeout(context.Background, time.Millisecond)
promise := client.GenerateAsync(ctx, 1)
// other stuff....
hash, err := promise.Receive()
// check error...
```

Reply:
The doc for context says not to store it in a struct. It's ok?

Reply:
I think it is proper in this case.  See https://github.com/golang/go/issues/22602.

Reply:
I see. Thank you.

How does it look now?